man/mosml.1
.TH MOSML 1 "18 September 2001" "Version 2.0"
.SH NAME
mosml \- Standard ML interpreter
.SH SYNOPSIS
.B mosml [\fIOPTION\fR]... [\fIFILE\fR]...
.SH DESCRIPTION
.PP
Invokes the interactive Moscow ML system. The interactive system allows
you to enter declarations and evaluate expressions.
.PP
You can quit the interactive session by typing \fBquit();\fR or
control-D.  Type \fBhelp "lib";\fR for an overview of built-in
function libraries, and e.g. \fBhelp "Array";\fR for help on
\fBArray\fR operations.
.PP
Invoking the interactive system with a list of files \fIFILE\fRs
correspond to invoking it without the file list, and when the
Moscow ML has started entering
\fBuse "file1";\fR ... \fBuse "filen";\fR.
.PP
This manual page documents the options accepted by 
.B mosml.
For more information consult the online manuals.
.SH OPTIONS
.TP
\fB\-conservative\fR
Sets conservative mode for compilation of subsequent units: accept
all extensions to the SML Modules language, but issue a warning for
each use. This is the default.
.TP
\fB\-I\fR \fIdirectory\fR
Specifies directories to be searched for interface files, bytecode
files, and source files.  A call to \fBuse\fR, \fBload\fR or
\fBloadOne\fR will first search the current directory, then all
directories specified by option \fB\-I\fR in order of appearance
from left to right, and finally the standard library directory.
(This option affects the variable \fBMeta.loadPath\fR.)
.TP
\fB\-imptypes\fR
Specifies that the type checker should distinguish between
imperative and applicative type variables, generalize all
applicative type variables, and generalize imperative type variables
only in non-expansive expressions.
.TP
\fB\-liberal\fR
Sets liberal mode for compilation of subsequent units: accept
without warnings all extensions to the SML Modules language.
.TP
\fB\-orthodox\fR
Sets orthodox mode for the compilation of subsequent units: reject
all uses of the extensions to the SML Modules language.  That is,
accept only SML Modules syntax.
.TP
\fB\-P\fR \fIunit-set\fR
Determines which library units will be included and open at
compile-time.  Any library unit in the load path can be used by the
\fBcompile\fR function for type checking purposes.  Thus regardless
\fB\-P\fR option, the \fBcompile\fR function knows the type of
library functions such as \fBArray.foldl\fR.
.RS
.TP
\fB\-P\fR \fIdefault\fR
The initial environment for the SML Basis Library: modules
\fBArray\fR, \fBChar\fR, \fBList\fR, \fBString\fR, and \fBVector\fR
will be loaded, and \fBChar\fR, \fBList\fR, and \fBString\fR will be
partially opened.
.TP
\fB\-P sml90\fR 
This provides an initial environment which is upwards compatible with
that of the 1990 `Definition of Standard ML' and with pre-1.30 releases
of Moscow ML. In particular, the functions \fBchr\fR, \fBexplode\fR,
\fBimplode\fR, and \fBord\fR work on strings, not characters.  The new
versions of these functions are still available as \fBChar.chr\fR,
\fBChar.ord\fR, \fBString.explode\fR, and \fBString.implode\fR.  The
math functions and input-output facilities required by the 1990
Definition are available at top-level.  In addition the same libraries
are loaded as with \fB-P default\fR.
.TP
\fB\-P nj93\fR
This provides a top-level environment which is mostly compatible with
that of SML/NJ 0.93.  The functions \fBapp\fR, \fBceiling\fR,
\fBchr\fR, \fBdec\fR, \fBexplode\fR, \fBfold\fR, \fBhd\fR, \fBimplode\fR,
\fBinc\fR, \fBmax\fR, \fBmin\fR, \fBnth\fR, \fBnthtail\fR, \fBord\fR,
\fBordof\fR, \fBrevapp\fR, \fBrevfold\fR, \fBsubstring\fR, \fBtl\fR,
and \fBtruncate\fR have the same type and meaning as in SML/NJ 0.93.
Note that this is incompatible with SML/NJ version 110. The math
functions and input-output facilities required by the 1990 `Definition
of Standard ML' are available at top-level.  In addition the same (new)
libraries are loaded as with \fB-P default\fR.  This option does not
imply \fB-imptypes\fR.
.TP
\fB\-P full\fR
This loads all the libraries marked \fBF\fR in the library list (see
the online manuals), and partially opens the \fBChar\fR, \fBList\fR,
and \fBString\fR units.
.TP
\fB\-P none\fR
No library units are loaded or opened initially.
.PP
Additional library units can loaded into the interactive system by
using the \fBload\fR function.
.RE
.TP
\fB\-quietdec\fR
Turns off the interactive system's prompt and responses, except for
warnings and error messages.  Useful for writing scripts in SML.
Sets \fBMeta.quietdec\fR to \fBtrue\fR.
.TP
\fB\-stdlib\fR \fIstdlib-directory\fR
Specify the standard library directory to be \fIstdlib-directory\fR. 
The default standard library is \fB/usr/lib/mosml\fR under Debian.
.TP
\fB\-valuepoly\fR
Specifies that the type checker should use `value polymorphism'.
Default.
.SH AUTHOR
Compiled by Henning Niss from the various Moscow ML documents 
written by Sergei Romanenko, Claudio Russo, and Peter Sestoft.
.SH "SEE ALSO"
.BR mosmlc (1),
.BR mosmlyac (1),
.BR mosmllex (1)
.PP
The full documentation for
.B mosml
and the other Moscow ML tools can be found in the online manuals in 
\fB/usr/share/doc/mosml/manual.pdf.gz\fR
(`Moscow ML Owner's Manual') and
\fB/usr/share/doc/mosml/mosmllib.pdf.gz\fR
(`Moscow ML Library Documentation').


CHANGELOG.md
# Unreleased

* SML Basis Library changes:
  * Added the structures `PackRealBig` and `PackRealLittle`
  * Added stop-gap implementation of the structure `Int64` and
    `Word64`. Stop-gap meaning that they don't fully follow the Basis
    Library specification, in particular there is no support for
    overloaded constant literals nor overloaded operations.

* Fixed some bugs in the configure part of the build systems, which
  meant that some features was sometimes disabled

* Updated the various test-suites so that they can be used to easily
  check if the compiler and mosmllib works.


# Version 2.10.1 (August 2014)

* ML Server Pages added to examples

* Build system cleaned up a bit


# Version 2.10 (August 2013)

* Support for dynamic linking and callback from C also under MacOS X

* There may now be 2 G globals (string literals etc) instead of 64 K

* Bugs fixed:
  * The HTML files generated for mosmllib now have valid URLs
  * `Substring.app` more efficient
  * Recursive structure compilation (elabRecSigExp) fixed
  * Several errors in `Mosmlcookie`
  * Misplaced double quote in `Msp.ahrefa`
  * Exception `Option` wasn't available at top-level
  * Linker did not check stamps of linked-in units
  * Hash function now much faster on very long strings
  * Double alignment constraints now correct with gcc-3.2 and Solaris
  * Bug in `Polyhash.filter`
  * `Socket` library returned only first 16 bytes of UDP datagram

* SML Basis Library changes:
  * `String` and `Substring`: added `concatWith`, `isSuffix`,
    `isSubstring`, `full`
  * `ListPair`: added `zipEq`, `appEq`, `mapEq`, `foldlEq`, `foldrEq`,
    `allEq`
  * Added `ArraySlice` and `VectorSlice` structures, also for `Char` and `Word8`
  * Added `find`, `findi`, `all`, `exists`, `collate` to `vector`, `array` and
    `slice` structures
  * Added `collate` to `List`
  * Added `update` to vector structures
  * `Path.{mkRelative,mkAbsolute}` now take record arguments
  * Negative `Time.time` values allowed
  * `OS.Process`: new functions `sleep` and `isSuccess`; `status` not `eqtype`
  * `Listsort`: added `eqclasses`, `merge`, `mergeUniq`
  * `Path`: added functions `isRoot`, `fromUnixPath`, `toUnixPath`, and
    exception `InvalidArc` (bug report by Henning Niss)
  * `TextIO`: `inputLine` now have type `instream -> string option`
  * `Array`: added type abbreviation `vector` (bug report by Andrzej
    Wasowski)
  * `FileSys`: type `access` renamed to `access_mode` (bug report by
    Henning Niss)
  * `Byte`: `unpackString{,Vec}` now uses vector slices (bug report by
    Henning Niss)
  * `Unix`: added functions `fromStatus`, `textInstreamOf`, `binInstreamOf`,
    `textOutstreamOf`, `binOutstreamOf`, `exit`, and added phantom types
    to proc (bug report by Henning Niss)
  * `Timer`: added function `checkCPUTimes`
  * `Word` and `Word8`: added functions `toLarge`, `toLargeX`, and `fromLarge`
    (bug report by Martin Elsman)
  * `General`: added exception Span and made the type of the function
    `before` less general (bug report by Henning Niss)

* Added modules to mosmllib:
  * `Buffer`: mutable string buffers for fast and efficient
    concatenation of strings
  * `Hashset`: sets implemented by hash-tables
  * `Rbset`: ordered sets implemented by red-black trees
  * `Redblackmap`: maps implemented by red-black trees

* Moscow ML is now developed in the open at github.


# Version 2.00 (June 2000)

* The full SML Modules language (structures, signatures, and functors)
  is now supported, thanks to Claudio Russo.  Also, several extensions
  to the SML Modules language are provided:
      - higher-order functors: functors may be defined within structures
        and functors
      - first-class modules: structures and functors may be packed and
        then handled as Core language values, which may then be unpacked
        as structures or functors again
      - recursive modules: signatures and structures may be recursively
        defined

* Value polymorphism has become friendlier: non-generalizable free
  type variables are left free, and become instantiated (once only)
  when the bound variable is used

* Added facilities for creating and communicating with subprocesses
  (structure `Unix` and `Signal` from SML Basis Library).

* Added facilities for efficient functional generation of HTML code
  (structure `Msp`); also supports the writing of ML Server Page scripts.

* Added facilities setting and accessing 'cookies' in CGI scripts
  (structure `Mosmlcookie`), thanks to Hans Molin, Uppsala, Sweden.

* The `Gdimage` structure now produces PNG images (using Thomas
  Boutell's gd library).


# Version 1.44 (August 1999)

## Added

* Access to GNU gdbm persistent hashtables (structures `Gdbm`, `Polygdbm`)

* Interface to the PostgreSQL database server (structure `Postgres`)

* Interface to the MySQL database server (structure `Mysql`)

* Interface to POSIX 1003.2 regular expressions (structure `Regex`)

* Interface to sockets (structure `Socket`)

* Faster bytecode execution (when compiled with GCC or egcs)

* Registration of ML and C functions simplify callbacks


src/launch.w32/mosml.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "driver.h"

__declspec(dllimport) int caml_main(int argc, char * argv[]);

char* stdlib = NULL;
char* opt = "";
char* usefiles = "";
char* mosmltop = NULL;

int argc2;
char** argv2;

int main(int argc, char* argv[])
{
  int i;
  char * a;

  stdlib = getenv("MOSMLLIB");
  if (stdlib == NULL) 
  {
    fprintf(stderr, "Variable MOSMLLIB is undefined.\n");
    exit(2);
  }

  // argv2 will be filled as follows:
  //     argv[0], %MOSMLLIB%\mosmltop, -stdlib, %MOSMLLIB%, options, filenames

  argv2 = (char **) xmalloc( sizeof(char *) * (argc + 4) );

  argv2[0] = argv[0];
  argv2[2] = "-stdlib";
  argc2 = 4;

  for (i = 1; i < argc; i++) 
  {
    a = argv[i];
    if (eq(a, "-I") || eq(a, "-include") || eq(a, "-P") || eq(a, "-perv") ||
        eq(a, "-m") || eq(a, "-msgstyle")) 
    {
      argv2[argc2++] = a;
      i++;
      if( i < argc )
        argv2[argc2++] = argv[i];
    }
    else if (eq(a, "-quietdec") || eq(a, "-valuepoly") || eq(a, "-imptypes") ||
             eq(a, "-orthodox") || eq(a, "-conservative") || eq(a, "-liberal") )
    {
      argv2[argc2++] = a;
    } 
    else if (eq(a, "-stdlib")) 
    {
      i++;
      if( i < argc )
        stdlib = argv[i];
    } 
    else if (prefix(a, "-")) 
    {
      fprintf(stderr, "Unknown option \"%s\", ignored\n", a);
    } 
    else 
    {
      /* reformat(&usefiles, "%s %s", usefiles, a); */
    }
  }

  for (i = 1; i < argc; i++) 
  {
    a = argv[i];
    if (eq(a, "-I") || eq(a, "-include") || eq(a, "-P") || eq(a, "-perv") ||
        eq(a, "-m") || eq(a, "-msgstyle")) 
      i++;
    else if (eq(a, "-quietdec") || eq(a, "-valuepoly") || eq(a, "-imptypes") ||
             eq(a, "-orthodox") || eq(a, "-conservative") || eq(a, "-liberal") )
      ; 
    else if (eq(a, "-stdlib")) 
      i++;
    else if (prefix(a, "-")) 
      ;
    else 
    {
      argv2[argc2++] = a;
    }
  }


  format(&mosmltop, "%s\\mosmltop", stdlib);
  argv2[1] = mosmltop;
  argv2[3] = stdlib;
  argv2[argc2] = NULL;

  return caml_main(argc2, argv2);
}
